lib/portlayer/storage/vsphere/image.go

```diff
@@ -283,6 +244,119 @@ func (v *ImageStore) WriteImage(ctx context.Context, parent *portlayer.Image, ID
 	return newImage, nil
 }
 
+// Create a temporary directory, create a vmdk in this directory, attach/mount
+// the disk, unpack the tar, check the checksum.  If the data doesn't match the
+// expected checksum, abort by nuking the tempdir.  If everything matches, move
+// the tmpdir to the expected location (with the vmdk inside it).  The unwind
+// path is a bit convoluted here;  we need to clean up on the way out in the
+// error case (using the tmpdir).
+func (v *ImageStore) writeImage(ctx context.Context, storeName, parentID, ID,
+	sum string, r io.Reader) error {
+
+	// Create the image directory in the store.
+	tmpImageDir := v.imageDirPath(storeName, ID+".inprogress")
```

Comment:
Docker uses [this](https://github.com/docker/docker/tree/master/distribution/xfer) to basically share the progress across multiple clients during concurrent download of the same layer(s). It has a collection of layers that are in flight, as well as a collection of `watcher`s that are concerned about one or more layers in flight. When a layer's progress gets updated, that `Progress` update struct is sent to each `watcher` watching that particular layer via a channel. 

We will definitely want something like this in the future.